### PR TITLE
Show PurchaseUrl if present in manifest

### PR DESCRIPTION
**Pull request by Trenly**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Are you working against an Issue?

This does not take into account any of the entitlement checking for the MSStore source; The only change this should have is to Pre-Indexed and Rest sources where, if the PurchaseUrl is present in the manifest, the URL will be displayed.

As part of this I have also taken the many if statements for checking if fields are empty and moved them into a function for ease of readability.

Loosely related to #1345

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2416)

### CODE REVIEW

File: src/AppInstallerCLICore/Workflows/ShowFlow.cpp [hunk 1]
```diff
@@ -12,8 +12,28 @@ using namespace AppInstaller::Utility;
 using namespace AppInstaller::Utility::literals;
 
 namespace {
-    void ShowMultiLineField(Execution::OutputStream& outputStream, AppInstaller::StringResource::StringId label, std::string& value)
+
+    template <typename CharT, typename String>
+    void ShowSingleLineField(Execution::OutputStream& outputStream, CharT label, String value, bool indent = false)
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/ShowFlow.cpp:17]: I didn't see anything but `StringID` values for label (and I wouldn't expect anything else).  I don't think it needs to be a templated type.  Applies to `ShowMultiLineField` as well.

And it's best to use `const String& value` to show the intent (we do not want to modify or own this parameter).

File: src/AppInstallerCLICore/Workflows/ShowFlow.cpp [hunk 2]
```diff
@@ -38,111 +72,49 @@ namespace AppInstaller::CLI::Workflow
...
-            {
-                info << "  "_liv << tag << std::endl;
-            }
-        }
+        ShowMultiValueField(info, Resource::String::ShowLabelTags, manifest.CurrentLocalization.Get<Manifest::Localization::Tags>());
         const auto& agreements = manifest.CurrentLocalization.Get<Manifest::Localization::Agreements>();
         if (!agreements.empty())
         {
-            context.Reporter.Info() << Execution::ManifestInfoEmphasis << Resource::String::ShowLabelAgreements << std::endl;
+            ShowSingleLineField(info, Resource::String::ShowLabelAgreements, " "_liv);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/ShowFlow.cpp:117]: Not a fan of reusing the function for something that is not semantically correct.

File: src/AppInstallerCLICore/Workflows/ShowFlow.cpp [hunk 3]
```diff
@@ -170,30 +142,15 @@ namespace AppInstaller::CLI::Workflow
...
-                info << "  "_liv << Execution::ManifestInfoEmphasis << Resource::String::ShowLabelInstallerProductId << ' ' << installer->ProductId << std::endl;
-            }
-            if (!installer->ReleaseDate.empty())
-            {
-                info << "  "_liv << Execution::ManifestInfoEmphasis << Resource::String::ShowLabelInstallerReleaseDate << ' ' << installer->ReleaseDate << std::endl;
-            }
+            ShowSingleLineField(info, Resource::String::ShowLabelInstallerType, Manifest::InstallerTypeToString(installer->InstallerType), true);
+            ShowSingleLineField(info, Resource::String::ShowLabelInstallerLocale, installer->Locale, true);
+            ShowSingleLineField(info, Resource::String::ShowLabelInstallerUrl, installer->Url, true);
+            ShowSingleLineField(info, Resource::String::ShowLabelInstallerSha256, (installer->InstallerType == Manifest::InstallerTypeEnum::MSStore) ? "" : Utility::SHA256::ConvertToString(installer->Sha256), true);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Workflows/ShowFlow.cpp:151]: The check should be based on the value at hand, not inferred from the installer type.

**Review by ryfu-msft — APPROVED** [review]

**Comment by ryfu-msft** [comment]
Looks good to me, will defer to @yao-msft for sign-off and merge

File: src/AppInstallerCLICore/Workflows/ShowFlow.cpp [hunk 4]
```diff
@@ -12,17 +12,57 @@ using namespace AppInstaller::Utility;
 using namespace AppInstaller::Utility::literals;
 
 namespace {
-    void ShowMultiLineField(Execution::OutputStream& outputStream, AppInstaller::StringResource::StringId label, std::string& value)
+
+    template <typename String>
+    void ShowSingleLineField(Execution::OutputStream outputStream, const AppInstaller::StringResource::StringId& label, const String& value, bool indent = false)
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/ShowFlow.cpp:17]: nit: StringId is actually a string_view, they'd better be passed by value.

Same for the MultiLine and multivalue one

File: src/AppInstallerCLICore/Workflows/ShowFlow.cpp [hunk 5]
```diff
@@ -12,17 +12,57 @@ using namespace AppInstaller::Utility;
...
+        if (value.empty())
+        {
+            return;
+        }
+        /*
+            We need to be able to find and replace within the string.However, we don't want to own the original string
+            Therefore, a copy is created here so we can manipulate it. The memory should be freed again once this method
+            returns and the string is no longer in scope.
+        */
+        std::string m_value = value;
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/ShowFlow.cpp:42]: nit: m_value is not a class member, maybe name it shownValue or something better?

File: src/AppInstallerCLICore/Workflows/ShowFlow.cpp [hunk 6]
```diff
@@ -175,36 +154,18 @@ namespace AppInstaller::CLI::Workflow
...
-            if (!installer->Sha256.empty())
-            {
-                info << "  "_liv << Execution::ManifestInfoEmphasis << Resource::String::ShowLabelInstallerSha256 << ' ' << Utility::SHA256::ConvertToString(installer->Sha256) << std::endl;
-            }
-            if (!installer->ProductId.empty())
-            {
-                info << "  "_liv << Execution::ManifestInfoEmphasis << Resource::String::ShowLabelInstallerProductId << ' ' << installer->ProductId << std::endl;
-            }
-            if (!installer->ReleaseDate.empty())
+            std::stringstream shownInstallerType;
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/ShowFlow.cpp:157]: I'm not sure if a stringstream is necessary here, can it be just the std::string

**Pull request merged**